cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkUtils.java

```diff
@@ -189,33 +195,96 @@ public static URLClassLoader createSparkFrameworkClassLoader(ClassLoader parentC
+    } catch (IOException e) {
+      LOG.warn("Failed to locate the rewritten Spark Assembly JAR. Fallback to use the original jar.", e);
+    }
+    localizeResources.put(sparkAssemblyJar.getName(), new LocalizeResource(sparkAssemblyJar));
+
+    // Shallow copy all files under directory defined by $HADOOP_CONF_DIR
+    // If $HADOOP_CONF_DIR is not defined, use the location of "yarn-site.xml" to determine the directory
+    // This is part of workaround for CDAP-5019 (SPARK-13441).
+    File hadoopConfDir = null;
+    if (System.getenv().containsKey(ApplicationConstants.Environment.HADOOP_CONF_DIR.key())) {
+      hadoopConfDir = new File(System.getenv(ApplicationConstants.Environment.HADOOP_CONF_DIR.key()));
+    } else {
+      URL yarnSiteLocation = SparkUtils.class.getClassLoader().getResource("yarn-site.xml");
+      if (yarnSiteLocation != null) {
+        try {
+          hadoopConfDir = new File(yarnSiteLocation.toURI()).getParentFile();
+        } catch (URISyntaxException e) {
+          // Shouldn't happen
+          LOG.warn("Failed to derive HADOOP_CONF_DIR from yarn-site.xml");
+        }
+      }
+    }
+    if (hadoopConfDir != null && hadoopConfDir.isDirectory()) {
+      try {
+        final File targetFile = File.createTempFile(LOCALIZED_CONF_DIR, ".zip", tempDir);
+        BundleJarUtil.createArchive(hadoopConfDir, new OutputSupplier<ZipOutputStream>() {
+          @Override
+          public ZipOutputStream getOutput() throws IOException {
+            return new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(targetFile)));
+          }
+        });
+        localizeResources.put(LOCALIZED_CONF_DIR, new LocalizeResource(targetFile, true));
+      } catch (IOException e) {
+        LOG.warn("Failed to create archive from {}", hadoopConfDir, e);
+      }
+    }
+
+    return sparkAssemblyJar.getName();
+  }
+
+  /**
+   * Creates a {@code __spark_conf__.zip} in the local directory by zipping contains from the
+   * {@code __spark_conf__} directory, with the addition of the serialization of the Spark properties.
+   * This method works in conjunction with the {@link #prepareSparkResources(CConfiguration, File, Map)} method and
+   * should only be called from the the Spark client container process.
+   *
+   * @param sparkProperties the Spark properties to serialize
+   */
+  public static void createSparkConfZip(final Properties sparkProperties) {
+    try {
+      BundleJarUtil.createArchive(new File(LOCALIZED_CONF_DIR), new OutputSupplier<ZipOutputStream>() {
+        @Override
+        public ZipOutputStream getOutput() throws IOException {
+          ZipOutputStream zipOutput = new ZipOutputStream(new FileOutputStream(LOCALIZED_CONF_DIR_ZIP));
+          try {
+            zipOutput.putNextEntry(new ZipEntry(SPARK_CONF_FILE));
+            sparkProperties.store(zipOutput, "Spark configuration.");
+            return zipOutput;
+          } catch (IOException e) {
+            zipOutput.close();
```

Comment:
This is intended not to because the `OutputSupplier` needs to provide an opened `ZipOutputStream`. We only close it in case of IOException to prevent resource leak and the IOException is propagated, hence no output stream will be returned from the `OutputSupplier`